README.md
[![Release](https://img.shields.io/github/v/release/abdelrahmanmostafa21/flutter-cloudimage-360-view)](https://github.com/abdelrahmanmostafa21/flutter-cloudimage-360-view/releases)
[![pub package](https://img.shields.io/pub/v/cloudimage_360_view.svg)](https://pub.dartlang.org/packages/cloudimage_360_view)
[![Contributions welcome](https://img.shields.io/badge/contributions-welcome-orange.svg)](#contributing)
[![License](https://img.shields.io/badge/license-MIT-red.svg)](https://opensource.org/licenses/MIT)
[![Cloudimage](https://img.shields.io/badge/Powered%20with%20%E2%99%A5%20-abdelrahman_mostafa-6986fa.svg)](https://www.github.com/abdelrahmanmostafa21)



<p align="center">
		<img
			alt="cloudimage logo"
            width=300,
			src="https://user-images.githubusercontent.com/38880077/167283693-b7a05609-ebc3-4cad-92b7-e7258c5898ae.png">
	</a>
</p>


<h1 align="center">
   Flutter Cloudimage 360 View
</h1>




<p align="center">A simple, interactive resource that can be used to provide a virtual tour of your product.</p>

<p align="center">	
    <img
		alt="Shoes"
    height=200
		src="https://user-images.githubusercontent.com/38880077/167283782-a0688a5d-867f-4e4f-9583-60355f0352a4.gif">
	<img
		alt="AirBuds"
    height=200
		src="https://user-images.githubusercontent.com/38880077/167283723-3ae10acc-8732-46e8-8e75-96b6a4d18a99.gif">

        
</p>


## <a name="table_of_contents"></a>Table of contents

* [Example](#example)
* [Installation](#installation)
* [How To Use](#how-to-use)
* [Image Model](#image-model)
* [Options](#options)
* [Controller](#controller)
* [Best practices](#best-practices)
* [Contributing](#contributing)
* [License](#license)


## <a name="example"></a> Example
To build and run the example app:

1. Get Flutter [here](https://flutter.dev) if you don't already have it
2. Clone this repository.
3. `cd` into the repo folder.
4. run `flutter run-android` or `flutter run-ios` to build the app.
- Please note that a Mac with XCode is required to build for iOS

>Example

<img 
    alt="Example Image"
    height= 300
    src="https://user-images.githubusercontent.com/38880077/167283738-8f8d0fca-d202-4861-81a3-1f63c7670c79.gif">


## <a name="installation"></a>Installation

Add `cloudimage_360_view: 1.0.3` to your `pubspec.yaml` dependencies. 
>Import Package

```dart
import 'package:cloudimage_360_view/cloudimage_360_view.dart';
```
## <a name="how-to-use"></a> How To Use
>Simply create a `Ci360View` widget, and pass the required parameters:

* [Horizontal Model (xImageModel | __optional__ if (yAxis) provided](#image-model)
* [Vertical Model (yImageModel) | __optional__ if (xAxis) provided](#image-model)
* [Ci360Options (options)](#options)
```dart
Ci360View(
    xImageModel: Ci360ImageModel.horizontal(
        imageFolder: 'your image x folder - check Ci360ImageModel',
        imageName: (index) => '$index.jpg',
        imagesLength: 233,
    ),
    yImageModel: Ci360ImageModel.vertical(
        imageFolder: 'your image y folder - check Ci360ImageModel',
        imageName: (index) => 'imagename-y-$index.jpg',
        imagesLength: 36,
    ),
    options: Ci360Options(
        swipeSensitivity: 2,
        autoRotate: true,
        rotationCount: 5,
        allowSwipeToRotate: true,
        onImageChanged: (index, reason, axis) {},
    ),
)
```

## <a name="image-model"></a> Image Model 
###### Type: **class** | _required_
>Horizontal (X-Axis)
```dart
Ci360ImageModel.horizontal({
    required this.imageFolder,
    required this.imageName,
    required this.imagesLength,
  });
```
###### Type: **class** | _required_
>Vertical (Y-Axis)
```dart
Ci360ImageModel.vertical({
    required this.imageFolder,
    required this.imageName,
    required this.imagesLength,
  });
```
#### imageFolder
###### Type: **String** | _required_
Your images folder on your cloud/cdn.
Base Data Folder Url For the main image.
`i.e https://domain.com/images/360-tesla/`

#### imageName
###### Type: **Function** | _required_
The filename pattern for your 360 image in `x|y` Axis builder.
`String Function(int index)`
Must return a valid `String` path with the called `index`
which the library will call with a number between `1` and `imageLength` for axis `x|y`.
##### `i.e`
```dart
Ci360Options.horizontal({
    imageName: (index) => 'tesla-x-$index.jpg'
  });
```
```dart
Ci360Options.vertical({
    imageName: (index) => 'tesla-y-$index.jpg'
  });
```
#### imagesLength
###### Type: **int** | _required_
Amount of images to load in `x | y` axis for 360 view.
Must Be Valid Length To Get The Valid FileName Of The Image

## <a name="options"></a> Options 
###### Type: **class** | _required_

```dart
Ci360Options({
    this.autoRotate = false,
    this.allowSwipeToRotate = true,
    this.rotationCount = 1,
    this.swipeSensitivity = 3,
    this.frameChangeDuration = kShortDuration,
    this.rotationDirection = CIRotationDirection.clockwise,
    this.onImageChanged,
  });
```

#### autoRotate
###### Type: **bool** | Default: **false** | _optional_
If set to true, the images will be displayed in incremented manner.

#### allowSwipeToRotate
###### Type: **bool** | Default: **true** | _optional_
If set to false, the gestures to rotate the image will be disabed.

#### rotationCount
###### Type: **int** | Default: **1** | _optional_
The number of cycles the whole image rotation should take place.
`0` means infinite rotation.

#### swipeSensitivity
###### Type: **int** | Default: **3** | _optional_

Based on the value the sensitivity of swipe gesture increases and decreases proportionally.
`1` slow and increases speed by `+1`

#### `frameChangeDuration`
###### Type: **Duration** | Default: **Duration(milliseconds: 80)** | _optional_
The time interval between shifting from one image frame to other.

#### rotationDirection
###### Type: **CIRotationDirection** | Default: **CIRotationDirection.clockwise** | _optional_
Based on the value the direction of rotation is set.

#### onImageChanged
###### Type: **Function** | Default: **null** | _optional_
Callback `Function(int index, CIImageChangedReason reason, Axis axis)` to provide you the index of current image when image frame is changed with the image axis and reason.

## <a name="controller"></a> Controller

### Ci360Controller

###### Type: **class** | Default: **"Ci360ControllerImpl"** | _optional_

The controller of the ci360 package.
You can implement your own controllers
##### `i.e`
```dart
class CustomCi360Controller implements Ci360Controller {
 // Should Override this variables as below
 @override
  Completer readyCompleter = Completer();

  Ci360State? _state;

  @override
  set state(Ci360State? state) {
    _state = state;
    if (!readyCompleter.isCompleted) {
      readyCompleter.complete();
    }
  }
// Implement the required methods with your custom callbacks
// Then Pass Your Custom Controller To [Ci360View] widget.
}
```
>Methods
### nextImage
###### Type: **Function**
manually roate to next image.
### previousImage
###### Type: **Function**
manually roate to previous image.

### rotateToImage
###### Type: **Function** | Parameter: **( int index, [Axis axis = Axis.horizontal] )**
manually roate to specified index on specified axis.

### stopAutoPlay
###### Type: **Function**
manually stop auto rotate of images.
This is a more on-demand way of doing this. Use the `autoRotate`
parameter in `Ci360Options` to specify the autoRotate behaviour of the ImageView.

### startAutoPlay
###### Type: **Function**
manually start auto rotate of images.
The ImageView will only autoPlay if the `autoRotate` parameter
in `Ci360Options` is true.


## <a name="best-practices"/> Best practices
In order to use cloudimage 360 view, your original (master) images should be stored on a server
or storage bucket (S3, Google Cloud, Azure Blob...) reachable over
HTTP or HTTPS.


## <a name="contributing"></a>Contributing!

All contributions are super welcome!

## <a name="license"></a>License
Flutter Cloudimage 360 View is provided under the [MIT License](https://opensource.org/licenses/MIT)


CHANGELOG.md
## v1.0.3
* update dart analysis options
* refactor some lines
* update README
## v1.0.2
* update README
* add local screenshots
* refactor folder names according to dart analyzer
## v1.0.1
* Update README
## v1.0.0
* Initial version.


example/lib/main.dart
import 'package:cloudimage_360_view/cloudimage_360_view.dart';
import 'package:example/components/index.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Cloud Image 360 View',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const ExamplePage(),
    );
  }
}

class ExamplePage extends StatefulWidget {
  const ExamplePage({Key? key}) : super(key: key);

  @override
  State<ExamplePage> createState() => _ExamplePageState();
}

class _ExamplePageState extends State<ExamplePage> {
  bool autoRotate = true;
  bool allowSwipe = true;
  int swipeSensitivity = 1;
  int rotationCount = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(8),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Image.asset(
                'assets/ci_360.png',
                height: 100,
              ),
              Text(
                'CloudImage 360(3D) ImageView',
                style: Theme.of(context)
                    .textTheme
                    .titleLarge
                    ?.copyWith(color: const Color.fromARGB(255, 2, 79, 124)),
                textAlign: TextAlign.center,
              ),
              smallVerticalSpacer,
              Text(
                'Engage your customers with a stunning 360 view of your products',
                style: Theme.of(context).textTheme.bodyLarge,
                textAlign: TextAlign.center,
              ),
              verticalSpacer,
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: AppListCard(
                  title: ExpansionTile(
                    title: const Text('Options'),
                    children: [
                      const Divider(),
                      AppSwitchTile(
                        title: const Text('Auto Rotate'),
                        value: autoRotate,
                        onChanged: (_) {
                          safeSetState(() {
                            autoRotate = !autoRotate;
                          });
                        },
                      ),
                      AppSwitchTile(
                        title: const Text('Allow Drag/Swipe'),
                        value: allowSwipe,
                        onChanged: (_) {
                          safeSetState(() {
                            allowSwipe = !allowSwipe;
                          });
                        },
                      ),
                      AppListTile(
                        title: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            const Text('Drag Sensetivity'),
                            Text(
                              'Faster...Slower',
                              style: Theme.of(context).textTheme.labelSmall,
                            ),
                          ],
                        ),
                        subtitle: SizedBox(
                          height: 35,
                          child: SingleChildScrollView(
                            padding: EdgeInsets.zero,
                            scrollDirection: Axis.horizontal,
                            child: Row(
                              children: [
                                for (var i = 1; i < 6; i++)
                                  ChoiceChip(
                                    selected: swipeSensitivity == i,
                                    label: Text('$i'),
                                    selectedColor: Colors.blueAccent.withOpacity(0.8),
                                    disabledColor: Colors.white54,
                                    onSelected: (_) {
                                      if (swipeSensitivity != i) {
                                        safeSetState(() {
                                          swipeSensitivity = i;
                                        });
                                      }
                                    },
                                  ),
                              ],
                            ),
                          ),
                        ),
                      ),
                      AppListTile(
                        title: const Text('Rotation Count'),
                        subtitle: SizedBox(
                          height: 35,
                          child: SingleChildScrollView(
                            padding: EdgeInsets.zero,
                            scrollDirection: Axis.horizontal,
                            child: Row(
                              children: [
                                for (var i = 0; i < 50; i++)
                                  ChoiceChip(
                                    selected: rotationCount == i,
                                    label: Text(i == 0 ? 'infinite' : '$i'),
                                    selectedColor: Colors.blueAccent.withOpacity(0.8),
                                    disabledColor: Colors.white54,
                                    onSelected: (_) {
                                      if (rotationCount != i) {
                                        safeSetState(() {
                                          rotationCount = i;
                                        });
                                      }
                                    },
                                  )
                              ],
                            ),
                          ),
                        ),
                      ),
                      verticalSpacer,
                      Text(
                        'more of features soon',
                        style:
                            Theme.of(context).textTheme.subtitle1?.copyWith(color: Colors.blueGrey),
                      ),
                      verticalSpacer,
                    ],
                  ),
                ),
              ),
              verticalSpacer,
              verticalSpacer,
              _TileBuilder(
                headerTitle: 'Horizontal & Vertical (x&y) Axis Eaxmple',
                child: Ci360View(
                  xImageModel: Ci360ImageModel.horizontal(
                    imageFolder: 'https://scaleflex.cloudimg.io/v7/demo/360-nike/',
                    imageName: (index) => 'nike-$index.jpg',
                    imagesLength: 35,
                  ),
                  yImageModel: Ci360ImageModel.vertical(
                    imageFolder: 'https://scaleflex.cloudimg.io/v7/demo/360-nike/',
                    imageName: (index) => 'nike-y-$index.jpg',
                    imagesLength: 36,
                  ),
                  options: Ci360Options(
                    swipeSensitivity: swipeSensitivity,
                    autoRotate: autoRotate,
                    rotationCount: rotationCount,
                    allowSwipeToRotate: allowSwipe,
                    onImageChanged: (index, reason, axis) {},
                  ),
                ),
              ),
              verticalSpacer,
              const Divider(thickness: 8),
              verticalSpacer,
              _TileBuilder(
                headerTitle: 'Horizontal(x) Axis Only Eaxmple',
                child: Ci360View(
                  xImageModel: Ci360ImageModel.horizontal(
                    imageFolder: 'https://scaleflex.cloudimg.io/v7/demo/earbuds/',
                    imageName: (index) => '$index.jpg',
                    imagesLength: 233,
                  ),
                  options: Ci360Options(
                    swipeSensitivity: swipeSensitivity,
                    autoRotate: autoRotate,
                    rotationCount: rotationCount,
                    allowSwipeToRotate: allowSwipe,
                    onImageChanged: (index, reason, axis) {},
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _TileBuilder extends StatelessWidget {
  const _TileBuilder({
    required this.headerTitle,
    required this.child,
    Key? key,
  }) : super(key: key);

  final String headerTitle;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8),
      child: Column(
        children: [
          Text(
            headerTitle,
            style: Theme.of(context).textTheme.bodyLarge,
          ),
          smallVerticalSpacer,
          child,
        ],
      ),
    );
  }
}
